libs/sdk-backend-tiger/src/convertors/toBackend/afm/TotalsConverter.ts

```diff
@@ -0,0 +1,114 @@
+        const includedDimensionIndexes = allDimensionIndexes.filter((idx) => dimIdx != idx);
+        const measureIdentifiers = totalsOfType.map((total) => total.measureIdentifier);
+        const includedDimensions = includedDimensionIndexes.map((includedDimIdx) => {
+            const dim = def.dimensions[includedDimIdx];
+            const dimensionAttributesValues = dim.itemIdentifiers.includes(MeasureGroupIdentifier)
+                ? [{ itemIdentifier: MeasureGroupIdentifier, values: measureIdentifiers }]
+                : null;
+            // FIXME synchronize dimensionIdentifier naming with convertDimensions
+            return { dimensionIdentifier: `dim_${includedDimIdx}`, dimensionAttributesValues };
+        });
+        const totalType = totalsOfType[0].type;
+
+        return {
+            localIdentifier: totalLocalIdentifier(totalType, dimIdx),
+            function: convertTotalType(totalType as TotalType),
+            includedDimensions,
+        };
+    });
+}
+
+/**
+ * Traverse given dimensions and their total definitions, group those by total type and call totalProcessor on each
+ * of these groups, together with current dimension index (dimIdx) and total type index in that dimension (typeIdx).
+ *
+ * This function captures the contract between execution response and execution result. Both of these response
+ * structures are synchronized using local identifiers induced from dimIdx typeIdx. In the future the function should
+ * be replaced by explicit concept of total local identifiers both in ITotal and ITotalDescriptor. Currently, given
+ * the rather limit functionality of general totals API, it's needed/useful only in Tiger.
+ */
+export function withTotals<T>(
+    dimensions: IDimension[],
+    totalProcessor: (dimIdx: number, typeIdx: number, totalsOfType: ITotal[]) => T,
+): T[] {
+    validateTotals(dimensions);
+    return flatMap(dimensions, (dim, dimIdx) => {
+        const totalsByType = groupBy(dim.totals || [], (total) => total.type);
+        return Object.keys(totalsByType).map((type, typeIdx) =>
+            totalProcessor(dimIdx, typeIdx, totalsByType[type]),
+        );
+    });
+}
+
+export function totalLocalIdentifier(type: TotalType, dimIdx: number) {
+    return `total_${dimIdx}_${type}`;
+}
+
+function validateTotals(dimensions: IDimension[]) {
+    dimensions.forEach((dim) => {
+        const firstDimensionItem = dim.itemIdentifiers[0];
+        dim.totals?.forEach((totalDef) => {
+            if (totalDef.attributeIdentifier != firstDimensionItem) {
+                throw new Error("Tiger backend does not support subtotals.");
+            }
+        });
+        const oppositeDimensions = dimensions.filter((dim0) => dim != dim0);
+        // NOTE measureGroup is assumed to be correctly placed at at most one dimension (checked elsewhere)
+        if (!isEmpty(dim.totals) && !oppositeDimensions.some(containsMeasureGroup)) {
```

Comment:
how about short-circuiting early in the fun if dim has no totals?

Reply:
I remember now. Because you cannot easily "continue" in forEach. I will switch to classic for...of loop where it's possible.